src/containerapp/azext_containerapp/_up_utils.py

```diff
@@ -1058,31 +1321,124 @@ def _set_up_defaults(
             env_list = [e for e in list_managed_environments(cmd=cmd) if e["name"] == env.name]
         else:
             env_list = [e for e in list_managed_environments(cmd=cmd) if e["name"] == env.name and format_location(e["location"]) == format_location(location)]
         if len(env_list) == 1:
             resource_group.name = parse_resource_id(env_list[0]["id"])["resource_group"]
+            env.resource_type = MANAGED_ENVIRONMENT_RESOURCE_TYPE
         if len(env_list) > 1:
             raise ValidationError(
                 f"There are multiple environments with name {env.name} on the subscription. "
                 "Please specify which resource group your Containerapp environment is in."
             )    # get ACR details from --image, if possible
+
+    _infer_existing_connected_env(cmd, location, resource_group, env, custom_location)
+
+    _infer_existing_custom_location_or_extension(cmd, name, location, resource_group, env, custom_location, extension)
+
     _get_acr_from_image(cmd, app)
 
 
+# Try to get existed connected environment
+# 'not env.resource_type' is for command: -n <app_name> --environment <env_name>:
+# 1. Find managed env with env name, if found, use it. We have done this in the previous parts.
+# 2. If manged env is not found, the env.resource_type is None, looks for a connected env with env name, uses it if found. If mutiple connected envs are found, thow a error.
+# (env.is_connected_environment() and (not env.name or not resource_group.name or not env.custom_location_id)) is for commands:
+# -n <app_name> --connected-cluster-id
+# -n <app_name> --custom-location
+def _infer_existing_connected_env(
+        cmd,
+        location,
+        resource_group: "ResourceGroup",
+        env: "ContainerAppEnvironment",
+        custom_location: "CustomLocation",
+):
+    if not env.resource_type or (env.is_connected_environment() and (not env.name or not resource_group.name or not env.custom_location_id)):
+        connected_env_list = list_connected_environments(cmd=cmd, resource_group_name=resource_group.name)
+        env_list = []
+        for e in connected_env_list:
+            if env.name and env.name != e["name"]:
+                continue
+            if location and format_location(e["location"]) != location:
+                continue
+            if custom_location.name and e["extendedLocation"]["name"].lower() != custom_location.get_rid().lower():
+                continue
+            if custom_location.connected_cluster_id:
+                custom_location_from_env = get_custom_location(cmd=cmd, custom_location_id=e["extendedLocation"]["name"])
+                if custom_location_from_env is None or custom_location.connected_cluster_id.lower() != custom_location_from_env.host_resource_id.lower():
+                    continue
+            env_list.append(e)
+
+        if len(env_list) == 1:
+            resource_group.name = parse_resource_id(env_list[0]["id"])["resource_group"]
```

Comment:
even there is rg provided, you will override it?

Reply:
No, the env_list is list with 
`connected_env_list = list_connected_environments(cmd=cmd, resource_group_name=resource_group.name)`
I will remove this line to avoid the confusion.